File: lua/server/lobby.lua
---@class Lobby : Object
---@field lobby fk.Room
Lobby = class("Lobby")

fk.lobby_callback = {}
local db = fk.ServerInstance:getDatabase()

function Lobby:initialize(_lobby)
  self.lobby = _lobby
  self.lobby.callback = function(_self, command, jsonData)
    local cb = fk.lobby_callback[command]
    if (type(cb) == "function") then
      cb(jsonData)
    else
      print("Lobby error: Unknown command " .. command);
    end
  end
end

fk.lobby_callback["UpdateAvatar"] = function(jsonData)
  -- jsonData: [ int uid, string newavatar ]
  local data = json.decode(jsonData)
  local id, avatar = data[1], data[2]
  local sql = "UPDATE userinfo SET avatar='%s' WHERE id=%d;"
  Sql.exec(db, string.format(sql, avatar, id))
  local player = fk.ServerInstance:findPlayer(id)
  player:setAvatar(avatar)
  player:doNotify("UpdateAvatar", avatar)
end

fk.lobby_callback["UpdatePassword"] = function(jsonData)
  -- jsonData: [ int uid, string oldpassword, int newpassword ]
  local data = json.decode(jsonData)
  local id, old, new = data[1], data[2], data[3]
  local sql_find = "SELECT password FROM userinfo WHERE id=%d;"
  local sql_update = "UPDATE userinfo SET password='%s' WHERE id=%d;"

  local passed = false
  local result = Sql.exec_select(db, string.format(sql_find, id))
  passed = (result["password"][1] == sha256(old))
  if passed then
    Sql.exec(db, string.format(sql_update, sha256(new), id))
  end

  local player = fk.ServerInstance:findPlayer(tonumber(id))
  player:doNotify("UpdatePassword", passed and "1" or "0")
end

fk.lobby_callback["CreateRoom"] = function(jsonData)
  -- jsonData: [ int uid, string name, int capacity ]
  local data = json.decode(jsonData)
  local owner = fk.ServerInstance:findPlayer(tonumber(data[1]))
  local roomName = data[2]
  local capacity = data[3]
  fk.ServerInstance:createRoom(owner, roomName, capacity)
end

fk.lobby_callback["EnterRoom"] = function(jsonData)
  -- jsonData: [ int uid, int roomId ]
  local data = json.decode(jsonData)
  local player = fk.ServerInstance:findPlayer(tonumber(data[1]))
  local room = fk.ServerInstance:findRoom(tonumber(data[2]))
  room:addPlayer(player)
end

function CreateRoom(_room)
  LobbyInstance = Lobby:new(_room)
end

File: src/server/lobby.cpp
#include "server/lobby.h"
#include "server/server.h"
#include "server/serverplayer.h"
#include "core/util.h"

Lobby::Lobby(Server *server) {
  this->server = server;
  setParent(server);
}

void Lobby::addPlayer(ServerPlayer *player) {
  if (!player) return;

  players.append(player);
  player->setRoom(this);

  if (player->getState() == Player::Robot) {
    removePlayer(player);
    player->deleteLater();
  } else {
    player->doNotify("EnterLobby", "[]");
  }

  server->updateOnlineInfo();
}

void Lobby::removePlayer(ServerPlayer *player) {
  players.removeOne(player);
  server->updateOnlineInfo();
}

void Lobby::updateAvatar(ServerPlayer *sender, const QString &jsonData) {
  auto arr = String2Json(jsonData).array();
  auto avatar = arr[0].toString();

  if (CheckSqlString(avatar)) {
    auto sql = QString("UPDATE userinfo SET avatar='%1' WHERE id=%2;")
      .arg(avatar)
      .arg(sender->getId());
    ExecSQL(ServerInstance->getDatabase(), sql);
    sender->setAvatar(avatar);
    sender->doNotify("UpdateAvatar", avatar);
  }
}

void Lobby::updatePassword(ServerPlayer *sender, const QString &jsonData) {
  auto arr = String2Json(jsonData).array();
  auto oldpw = arr[0].toString();
  auto newpw = arr[1].toString();
  auto sql_find =
    QString("SELECT password, salt FROM userinfo WHERE id=%1;")
    .arg(sender->getId());

  auto passed = false;
  auto arr2 = SelectFromDatabase(ServerInstance->getDatabase(), sql_find);
  auto result = arr2[0].toObject();
  passed = (result["password"].toString() ==
      QCryptographicHash::hash(
        oldpw.append(result["salt"].toString()).toLatin1(),
        QCryptographicHash::Sha256)
      .toHex());
  if (passed) {
    auto sql_update =
      QString("UPDATE userinfo SET password='%1' WHERE id=%2;")
      .arg(QCryptographicHash::hash(
            newpw.append(result["salt"].toString()).toLatin1(),
            QCryptographicHash::Sha256)
          .toHex())
      .arg(sender->getId());
    ExecSQL(ServerInstance->getDatabase(), sql_update);
  }

  sender->doNotify("UpdatePassword", passed ? "1" : "0");
}

void Lobby::createRoom(ServerPlayer *sender, const QString &jsonData) {
  auto arr = String2Json(jsonData).array();
  auto name = arr[0].toString();
  auto capacity = arr[1].toInt();
  auto timeout = arr[2].toInt();
  auto settings =
    QJsonDocument(arr[3].toObject()).toJson(QJsonDocument::Compact);
  ServerInstance->createRoom(sender, name, capacity, timeout, settings);
}

void Lobby::getRoomConfig(ServerPlayer *sender, const QString &jsonData) {
  auto arr = String2Json(jsonData).array();
  auto roomId = arr[0].toInt();
  auto room = ServerInstance->findRoom(roomId);
  if (room) {
    auto settings = room->getSettings();
    // 手搓JSON数组 跳过编码解码
    sender->doNotify("GetRoomConfig", QString("[%1,%2]").arg(roomId).arg(settings));
  } else {
    sender->doNotify("ErrorMsg", "no such room");
  }
}

void Lobby::enterRoom(ServerPlayer *sender, const QString &jsonData) {
  auto arr = String2Json(jsonData).array();
  auto roomId = arr[0].toInt();
  auto room = ServerInstance->findRoom(roomId);
  if (room) {
    auto settings = QJsonDocument::fromJson(room->getSettings());
    auto password = settings["password"].toString();
    if (password.isEmpty() || arr[1].toString() == password) {
      if (room->isOutdated()) {
        sender->doNotify("ErrorMsg", "room is outdated");
      } else {
        room->addPlayer(sender);
      }
    } else {
      sender->doNotify("ErrorMsg", "room password error");
    }
  } else {
    sender->doNotify("ErrorMsg", "no such room");
  }
}

void Lobby::observeRoom(ServerPlayer *sender, const QString &jsonData) {
  auto arr = String2Json(jsonData).array();
  auto roomId = arr[0].toInt();
  auto room = ServerInstance->findRoom(roomId);
  if (room) {
    auto settings = QJsonDocument::fromJson(room->getSettings());
    auto password = settings["password"].toString();
    if (password.isEmpty() || arr[1].toString() == password) {
      if (room->isOutdated()) {
        sender->doNotify("ErrorMsg", "room is outdated");
      } else {
        room->addObserver(sender);
      }
    } else {
      sender->doNotify("ErrorMsg", "room password error");
    }
  } else {
    sender->doNotify("ErrorMsg", "no such room");
  }
}

void Lobby::refreshRoomList(ServerPlayer *sender, const QString &) {
  ServerInstance->updateRoomList(sender);
};

typedef void (Lobby::*room_cb)(ServerPlayer *, const QString &);

void Lobby::handlePacket(ServerPlayer *sender, const QString &command,
                        const QString &jsonData) {
  static const QMap<QString, room_cb> lobby_actions = {
    {"UpdateAvatar", &Lobby::updateAvatar},
    {"UpdatePassword", &Lobby::updatePassword},
    {"CreateRoom", &Lobby::createRoom},
    {"GetRoomConfig", &Lobby::getRoomConfig},
    {"EnterRoom", &Lobby::enterRoom},
    {"ObserveRoom", &Lobby::observeRoom},
    {"RefreshRoomList", &Lobby::refreshRoomList},
    {"Chat", &Lobby::chat},
  };

  auto func = lobby_actions[command];
  if (func) (this->*func)(sender, jsonData);
}
